src/main/rx/ghst.c

```diff
@@ -0,0 +1,259 @@
+
+    if (ghstFrameIdx < fullFrameLength) {
+        ghstFrame.bytes[ghstFrameIdx++] = (uint8_t)c;
+        if (ghstFrameIdx >= fullFrameLength) {                  // buffer is full?
+            ghstFrameIdx = 0;
+            const uint8_t crc = ghstFrameCRC();
+            if (crc == ghstFrame.bytes[fullFrameLength - 1]) {
+                switch (ghstFrame.frame.type)
+                {
+                    case GHST_UL_RC_CHANS_HS4_5TO8:
+                    case GHST_UL_RC_CHANS_HS4_9TO12:
+                    case GHST_UL_RC_CHANS_HS4_13TO16:
+                        if (ghstFrame.frame.addr == GHST_ADDR_FC) {
+                            lastRcFrameTimeUs = currentTimeUs;
+                            ghstFrameDone = true;
+                            memcpy(&ghstChannelDataFrame, &ghstFrame, sizeof(ghstFrame));
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+        }
+    }
+}
+STATIC_UNIT_TESTED uint8_t ghstFrameStatus(rxRuntimeState_t *rxRuntimeState)
+{
+    UNUSED(rxRuntimeState);
+    int iStartIdx = 4;
+
+    if (ghstFrameDone) {
+        ghstFrameDone = false;
+
+        const ghstPayloadPulses_t* const rcChannels = (ghstPayloadPulses_t*)&ghstChannelDataFrame.frame.payload;
+
+        // all uplink frames contain CH1..4 data (12 bit)
+        ghstChannelData[0] = rcChannels->ch1 >> 1;
+        ghstChannelData[1] = rcChannels->ch2 >> 1;
+        ghstChannelData[2] = rcChannels->ch3 >> 1;
+        ghstChannelData[3] = rcChannels->ch4 >> 1;
+
+        // remainder of uplink frame contains 4 more channels (8 bit), sent in a round-robin fashion
+        switch(ghstChannelDataFrame.frame.type)
+        {
+            case GHST_UL_RC_CHANS_HS4_5TO8:		iStartIdx = 4;	break;
+            case GHST_UL_RC_CHANS_HS4_9TO12:	iStartIdx = 8;	break;
+            case GHST_UL_RC_CHANS_HS4_13TO16:   iStartIdx = 12;	break;
+        } 
+
+        ghstChannelData[iStartIdx++] = rcChannels->cha << 3;
+        ghstChannelData[iStartIdx++] = rcChannels->chb << 3;
+        ghstChannelData[iStartIdx++] = rcChannels->chc << 3;
+        ghstChannelData[iStartIdx++] = rcChannels->chd << 3;
+
+        return RX_FRAME_COMPLETE;
+    }
+    return RX_FRAME_PENDING;
+}
+
+STATIC_UNIT_TESTED uint16_t ghstReadRawRC(const rxRuntimeState_t *rxRuntimeState, uint8_t chan)
+{
+    UNUSED(rxRuntimeState); 
+
+    // derived from original SBus scaling, with slight correction for offset (now symmetrical 
+    // around OpenTx 0 value)
+    // scaling is: 
+    //      OpenTx   RC     PWM
+    // min  -1024    172    988us
+    // ctr  0        992    1500us
+    // max  1024     1811   2012us   
+    //
+
+    return (5 * (ghstChannelData[chan]+1) / 8) + 880;
+}
+
+// called from telemetry/ghst.c
+void ghstRxWriteTelemetryData(const void *data, int len)
```

Comment:
This and the following could probably be moved into `telemetry/ghst.c` for simplification if `serialPort` is made accessible there.